src/Workspaces/Core/Portable/ExternalAccess/UnitTesting/UnitTestingIncrementalAnalyzerProvider.cs

```diff
@@ -10,11 +11,23 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.UnitTesting
     internal class UnitTestingIncrementalAnalyzerProvider : IIncrementalAnalyzerProvider
     {
         private readonly IUnitTestingIncrementalAnalyzerProviderImplementation _incrementalAnalyzerProvider;
+        private IIncrementalAnalyzer _analyzer;
 
         public UnitTestingIncrementalAnalyzerProvider(IUnitTestingIncrementalAnalyzerProviderImplementation incrementalAnalyzerProvider)
             => _incrementalAnalyzerProvider = incrementalAnalyzerProvider;
 
         public IIncrementalAnalyzer CreateIncrementalAnalyzer(Workspace workspace)
-            => new UnitTestingIncrementalAnalyzer(_incrementalAnalyzerProvider.CreateIncrementalAnalyzer(workspace));
+        {
+            // NOTE: We're currently expecting the analyzer to be singleton, so that
+            //       analyzers returned when calling this method twice would pass a reference equality check.
+            //       One instance should be created by SolutionCrawler, another one by us, when calling the
+            //       UnitTestingSolutionCrawlerServiceAccessor.Reanalyze method.
+            if (_analyzer == null)
+            {
+                _analyzer = new UnitTestingIncrementalAnalyzer(_incrementalAnalyzerProvider.CreateIncrementalAnalyzer());
```

Comment:
>_analyzer [](start = 16, length = 9)

Do you not need this to be thread-safe? I.e. using Lazy<T> for example?

Reply:
Or couldn't you move this to the constructor and create UnitTestingIncrementalAnalyzer there?

---
In reply to: [382820313](https://github.com/dotnet/roslyn/pull/41848#discussion_r382820313) [](ancestors = 382820313)

Reply:
Its not possible in the current implementation for this to be called concurrently. However, I agree we should make this thread safe (e.g. compare exchange) as this is supposed to be a singleton.